File: backend/app/schemas/blank_check.py
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class ErrorPayload(BaseModel):
    code: str = Field(..., description="Stable machine-readable error code")
    message: str = Field(..., description="Human-readable error message")
    details: dict[str, Any] | None = Field(
        default=None, description="Optional structured error details"
    )


class ErrorResponse(BaseModel):
    error: ErrorPayload


class BlankCheckRequest(BaseModel):
    page: int = Field(
        default=0,
        ge=0,
        description="Zero-based page index within the PDF. Optional in form-data.",
    )


class BlankCheckResponse(BaseModel):
    variant: list[str]
    date: list[str]
    reg_number: list[str]
    answers: list[list[str]]
    repl: list[list[str]]
    record_id: int
    warnings: list[str] = Field(default_factory=list)
    aligned_image_url: str | None = Field(
        default=None,
        description="URL для скачивания выровненного изображения страницы бланка",
    )
        

class IssueCode(str, Enum):
    MINUS_NOT_LEADING = "MINUS_NOT_LEADING"
    INTERNAL_EMPTY_CELL = "INTERNAL_EMPTY_CELL"
    LEADING_EMPTY_CELL = "LEADING_EMPTY_CELL"
    NOT_AN_INTEGER = "NOT_AN_INTEGER"
    MULTIPLE_MINUS = "MULTIPLE_MINUS"
    EMPTY_AFTER_TRIM = "EMPTY_AFTER_TRIM"
    UNSUPPORTED_SYMBOL = "UNSUPPORTED_SYMBOL"
    REQUIRED_FIELD_EMPTY = "REQUIRED_FIELD_EMPTY"


class RecognizedCell(BaseModel):
    index: int = Field(..., description="Zero-based index within the field")
    row: int | None = Field(
        default=None, description="Row index for grid-like fields (if applicable)"
    )
    col: int | None = Field(
        default=None, description="Column index for grid-like fields (if applicable)"
    )
    symbol: str | None = Field(
        default=None,
        description=(
            "Recognized symbol: '0'-'9', '-', 'E' (empty from OCR), "
            "'S' (unreadable), or null/empty when explicitly cleared"
        ),
    )


class ValidationIssue(BaseModel):
    field_id: str = Field(..., description="Identifier of the field with an issue")
    cell_indices: list[int] = Field(
        default_factory=list,
        description="Indices of cells related to this issue (within the field)",
    )
    code: IssueCode
    message: str


class FieldReview(BaseModel):
    field_id: str = Field(..., description="Stable identifier of the numeric field")
    label: str = Field(..., description="Human-friendly label for display in UI")
    cells: list[RecognizedCell]
    issues: list[ValidationIssue] = Field(default_factory=list)
    proposed_joined: str = Field(
        default="",
        description="Joined string built from non-empty symbols in order",
    )
    parsed_integer: int | None = Field(
        default=None, description="Parsed integer value if the field is valid"
    )
    is_valid: bool = Field(
        ..., description="True if the field passes all validation rules"
    )


class CorrectionFieldSubmission(BaseModel):
    field_id: str
    cells: list[RecognizedCell]
    joined_value: str | None = Field(
        default=None,
        description="Optional whole-number representation provided by the frontend",
    )


class CorrectionPayload(BaseModel):
    page: int = Field(..., description="Zero-based index of the processed PDF page")
    source_filename: str | None = Field(
        default=None,
        description="Original PDF filename (for multi-page upload, set when returning 422)",
    )
    aligned_image_url: str | None = Field(
        default=None, description="URL of the aligned page image to show in UI"
    )
    fields: list[FieldReview] = Field(
        ..., description="Per-field layout, symbols, and validation issues"
    )


class BlankListItem(BaseModel):
    """One item in the list of uploaded blanks (GET /v1/blanks)."""

    id: int
    source_filename: str | None = None
    source_url: str | None = None
    page_num: int | None = None
    created_at: str = Field(..., description="ISO datetime")
    verified: bool = Field(default=False, description="Whether the blank was manually verified")
    verified_at: str | None = Field(default=None, description="ISO datetime when verified")
    verified_by: str | None = Field(default=None, description="Login of user who verified")
    variant: list[str] = Field(default_factory=list)
    date: list[str] = Field(default_factory=list)
    reg_number: list[str] = Field(default_factory=list)


class BlankEditResponse(BaseModel):
    """Single blank for edit UI: CorrectionPayload + record_id (GET /v1/blanks/{id})."""

    record_id: int
    page: int
    aligned_image_url: str | None = None
    fields: list[FieldReview]
    verified: bool = Field(default=False, description="Whether the blank was manually verified")
    verified_at: str | None = Field(default=None, description="ISO datetime when verified")
    verified_by: str | None = Field(default=None, description="Login of user who verified")


class CorrectionSubmission(BaseModel):
    page: int = Field(..., description="Zero-based index of the corrected page")
    source_filename: str | None = Field(
        default=None,
        description="Original PDF filename (send when saving after multi-page correction)",
    )
    fields: list[CorrectionFieldSubmission]
    aligned_image_url: str | None = Field(
        default=None,
        description="URL of the aligned page image (S3), to store in source_url when saving",
    )
    record_id: int | None = Field(
        default=None,
        description="If set, update this existing record instead of creating a new one",
    )
    verified: bool | None = Field(
        default=None,
        description="If set when updating (record_id), set blank as verified/unverified (works for main admin too)",
    )


class SavedRecordIdItem(BaseModel):
    """One saved blank from multi-page processing."""

    page: int = Field(..., description="Zero-based page index")
    record_id: int = Field(..., description="Database record id")


class MultiPageErrorDetails(BaseModel):
    """422 details when some pages have validation errors (multi-page upload)."""

    pages_with_errors: list[CorrectionPayload] = Field(
        ...,
        description="Payloads for pages that require manual correction",
    )
    saved_record_ids: list[SavedRecordIdItem] = Field(
        default_factory=list,
        description="Pages that were saved successfully",
    )


class MultiPageSuccessResponse(BaseModel):
    """200 response when all pages were processed and saved (multi-page upload)."""

    saved_record_ids: list[SavedRecordIdItem] = Field(
        ...,
        description="All saved records, one per page",
    )


class SetVerifiedBody(BaseModel):
    """Body for PATCH /v1/blanks/{id}/verified."""

    verified: bool = Field(..., description="Set blank as verified (true) or unchecked (false)")

File: frontend/src/api/blankCheck.ts
export interface BlankCheckResult {
  variant: string[];
  date: string[];
  reg_number: string[];
  answers: string[][];
  repl: string[][];
  record_id: number;
  warnings: string[];
  aligned_image_url?: string | null;
}

import { authHeaders, handleUnauthorized } from "./auth";

export type IssueCode =
  | "MINUS_NOT_LEADING"
  | "INTERNAL_EMPTY_CELL"
  | "LEADING_EMPTY_CELL"
  | "NOT_AN_INTEGER"
  | "MULTIPLE_MINUS"
  | "EMPTY_AFTER_TRIM"
  | "UNSUPPORTED_SYMBOL"
  | "REQUIRED_FIELD_EMPTY";

export interface RecognizedCell {
  index: number;
  row?: number | null;
  col?: number | null;
  symbol: string | null;
}

export interface ValidationIssue {
  field_id: string;
  cell_indices: number[];
  code: IssueCode;
  message: string;
}

export interface FieldReview {
  field_id: string;
  label: string;
  cells: RecognizedCell[];
  issues: ValidationIssue[];
  proposed_joined: string;
  parsed_integer: number | null;
  is_valid: boolean;
}

export interface CorrectionPayload {
  page: number;
  source_filename?: string | null;
  aligned_image_url?: string | null;
  fields: FieldReview[];
}

/** One saved record from multi-page processing (200 or 422 response). */
export interface SavedRecordIdItem {
  page: number;
  record_id: number;
}

/** 200 response from POST /v1/blank-check/multi when all pages saved. */
export interface MultiPageSuccessResponse {
  saved_record_ids: SavedRecordIdItem[];
}

/** 422 details when some pages have errors (REVIEW_REQUIRED from multi endpoint). */
export interface MultiPageErrorDetails {
  pages_with_errors: CorrectionPayload[];
  saved_record_ids: SavedRecordIdItem[];
}

/** One item in the list from GET /v1/blanks */
export interface BlankListItem {
  id: number;
  source_filename?: string | null;
  source_url?: string | null;
  page_num?: number | null;
  created_at: string;
  verified?: boolean;
  verified_at?: string | null;
  verified_by?: string | null;
  variant: string[];
  date: string[];
  reg_number: string[];
}

/** Response from GET /v1/blanks/:id for edit UI */
export interface BlankEditResponse extends CorrectionPayload {
  record_id: number;
  verified?: boolean;
  verified_at?: string | null;
  verified_by?: string | null;
}

export interface CorrectionFieldSubmission {
  field_id: string;
  cells: RecognizedCell[];
  joined_value?: string | null;
}

export interface CorrectionSubmission {
  page: number;
  source_filename?: string | null;
  fields: CorrectionFieldSubmission[];
  aligned_image_url?: string | null;
  record_id?: number | null;
}

export interface ApiErrorPayload {
  code: string;
  message: string;
  details?: Record<string, unknown> | CorrectionPayload | MultiPageErrorDetails;
}

export interface ApiErrorResponse {
  error: ApiErrorPayload;
}

export class ApiError extends Error {
  readonly status: number;
  readonly code: string;
  readonly details?: Record<string, unknown> | CorrectionPayload | MultiPageErrorDetails;

  constructor(status: number, payload: ApiErrorPayload) {
    super(payload.message);
    this.status = status;
    this.code = payload.code;
    this.details = payload.details;
  }
}

export class NetworkError extends Error {
  constructor(message = "Не удалось подключиться к серверу") {
    super(message);
  }
}

export function isCorrectionPayload(
  value: unknown,
): value is CorrectionPayload {
  if (!value || typeof value !== "object") return false;
  const maybe = value as Partial<CorrectionPayload>;
  return (
    typeof maybe.page === "number" &&
    Array.isArray(maybe.fields)
  );
}

export function isMultiPageErrorDetails(
  value: unknown,
): value is MultiPageErrorDetails {
  if (!value || typeof value !== "object") return false;
  const maybe = value as Partial<MultiPageErrorDetails>;
  return (
    Array.isArray(maybe.pages_with_errors) &&
    Array.isArray(maybe.saved_record_ids)
  );
}

const API_BASE =
  typeof import.meta.env?.VITE_API_URL === "string"
    ? import.meta.env.VITE_API_URL
    : "";

export async function submitCorrections(
  submission: CorrectionSubmission,
  timeoutMs = 60000,
): Promise<BlankCheckResult> {
  const url = `${API_BASE}/api/v1/blank-check/corrections`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      headers: authHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify(submission),
      signal: controller.signal,
    });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  const text = await res.text();

  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  try {
    return JSON.parse(text) as BlankCheckResult;
  } catch {
    throw new ApiError(res.status, {
      code: "INVALID_RESPONSE",
      message: "Сервер вернул некорректный ответ",
      details: undefined,
    });
  }
}

export async function fetchBlanksList(
  search?: string,
  uncheckedOnly?: boolean,
  timeoutMs = 15000,
): Promise<BlankListItem[]> {
  const params = new URLSearchParams();
  if (search != null && search.trim() !== "") {
    params.set("search", search.trim());
  }
  if (uncheckedOnly === true) {
    params.set("unchecked_only", "true");
  }
  const query = params.toString();
  const url = `${API_BASE}/api/v1/blanks${query ? `?${query}` : ""}`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, { method: "GET", headers: authHeaders(), signal: controller.signal });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  const text = await res.text();
  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  try {
    return JSON.parse(text) as BlankListItem[];
  } catch {
    throw new ApiError(res.status, {
      code: "INVALID_RESPONSE",
      message: "Сервер вернул некорректный ответ",
      details: undefined,
    });
  }
}

export async function fetchBlankById(
  id: number,
  timeoutMs = 15000,
): Promise<BlankEditResponse> {
  const url = `${API_BASE}/api/v1/blanks/${id}`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, { method: "GET", headers: authHeaders(), signal: controller.signal });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  const text = await res.text();
  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  try {
    return JSON.parse(text) as BlankEditResponse;
  } catch {
    throw new ApiError(res.status, {
      code: "INVALID_RESPONSE",
      message: "Сервер вернул некорректный ответ",
      details: undefined,
    });
  }
}

export async function deleteBlankApi(
  id: number,
  timeoutMs = 15000,
): Promise<void> {
  const url = `${API_BASE}/api/v1/blanks/${id}`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  const res = await fetch(url, {
    method: "DELETE",
    headers: authHeaders(),
    signal: controller.signal,
  });
  clearTimeout(timer);
  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    const text = await res.text();
    try {
      const body = text ? (JSON.parse(text) as Record<string, unknown>) : {};
      const detail = body.detail;
      const msg = typeof detail === "string" ? detail : text || `HTTP ${res.status}`;
      throw new ApiError(res.status, { code: "HTTP_ERROR", message: msg, details: undefined });
    } catch (e) {
      if (e instanceof ApiError) throw e;
      throw new ApiError(res.status, {
        code: "HTTP_ERROR",
        message: text || `HTTP ${res.status}`,
        details: undefined,
      });
    }
  }
}

export async function setBlankVerifiedApi(
  id: number,
  verified: boolean,
  timeoutMs = 15000,
): Promise<void> {
  const url = `${API_BASE}/api/v1/blanks/${id}/verified`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  const res = await fetch(url, {
    method: "PATCH",
    headers: authHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify({ verified }),
    signal: controller.signal,
  });
  clearTimeout(timer);
  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    const text = await res.text();
    try {
      const body = text ? (JSON.parse(text) as Record<string, unknown>) : {};
      const detail = body.detail;
      const msg = typeof detail === "string" ? detail : text || `HTTP ${res.status}`;
      throw new ApiError(res.status, { code: "HTTP_ERROR", message: msg, details: undefined });
    } catch (e) {
      if (e instanceof ApiError) throw e;
      throw new ApiError(res.status, {
        code: "HTTP_ERROR",
        message: text || `HTTP ${res.status}`,
        details: undefined,
      });
    }
  }
}

export async function uploadPdfAndPredict(
  file: File,
  page: number = 0,
  timeoutMs = 60000
): Promise<BlankCheckResult> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("filename", file.name);
  formData.append("page", String(page));

  const url = `${API_BASE}/api/v1/blank-check`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      body: formData,
      headers: authHeaders(),
      signal: controller.signal,
    });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  const text = await res.text();

  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  try {
    return JSON.parse(text) as BlankCheckResult;
  } catch {
    throw new ApiError(res.status, {
      code: "INVALID_RESPONSE",
      message: "Сервер вернул некорректный ответ",
      details: undefined,
    });
  }
}

export async function uploadPdfAndPredictMulti(
  file: File,
  timeoutMs = 180000,
): Promise<MultiPageSuccessResponse> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("filename", file.name);

  const url = `${API_BASE}/api/v1/blank-check/multi`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      body: formData,
      headers: authHeaders(),
      signal: controller.signal,
    });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  const text = await res.text();

  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  try {
    return JSON.parse(text) as MultiPageSuccessResponse;
  } catch {
    throw new ApiError(res.status, {
      code: "INVALID_RESPONSE",
      message: "Сервер вернул некорректный ответ",
      details: undefined,
    });
  }
}

export async function downloadBlanksTable(
  timeoutMs = 60000,
  filename = "blanks.xlsx",
): Promise<void> {
  const url = `${API_BASE}/api/v1/export`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  let res: Response;
  try {
    res = await fetch(url, { method: "GET", headers: authHeaders(), signal: controller.signal });
  } catch (err) {
    clearTimeout(timer);
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new NetworkError("Превышено время ожидания ответа от сервера");
    }
    throw new NetworkError();
  } finally {
    clearTimeout(timer);
  }

  if (!res.ok) {
    if (res.status === 401) handleUnauthorized();
    const text = await res.text();
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      const rawDetail = body.detail;
      const nestedError =
        typeof rawDetail === "object" && rawDetail !== null && "error" in rawDetail
          ? (rawDetail as { error: ApiErrorPayload }).error
          : undefined;
      const errorPayload =
        (body.error as ApiErrorPayload) ??
        nestedError ??
        (rawDetail as ApiErrorPayload);
      if (
        errorPayload &&
        typeof errorPayload.code === "string" &&
        typeof errorPayload.message === "string"
      ) {
        throw new ApiError(res.status, errorPayload);
      }
    } catch (e) {
      if (e instanceof ApiError) throw e;
    }
    throw new ApiError(res.status, {
      code: "HTTP_ERROR",
      message: text || `HTTP ${res.status}`,
      details: undefined,
    });
  }

  const blob = await res.blob();
  const disposition = res.headers.get("Content-Disposition");
  const match = disposition?.match(/filename="?([^";\n]+)"?/);
  const name = match?.[1]?.trim() || filename;

  const a = document.createElement("a");
  a.href = URL.createObjectURL(blob);
  a.download = name;
  a.click();
  URL.revokeObjectURL(a.href);
}
